src/hasher512.cpp

```diff
@@ -0,0 +1,235 @@
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/hasher512.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/assert.hpp"
+
+#if TORRENT_USE_CRYPTOAPI
+namespace
+{
+	HCRYPTPROV make_crypt_provider()
+	{
+		using namespace libtorrent;
+
+		HCRYPTPROV ret;
+		if (CryptAcquireContext(&ret, nullptr, nullptr, PROV_RSA_AES
+			, CRYPT_VERIFYCONTEXT) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+		return ret;
+	}
+
+	HCRYPTPROV get_crypt_provider()
+	{
+		static HCRYPTPROV prov = make_crypt_provider();
+		return prov;
+	}
+}
+#endif
+
+namespace libtorrent
+{
+	hasher512::hasher512()
+	{
+#ifdef TORRENT_USE_LIBGCRYPT
+		gcry_md_open(&m_context, GCRY_MD_SHA512, 0);
+#elif TORRENT_USE_COMMONCRYPTO
+		CC_SHA512_Init(&m_context);
+#elif TORRENT_USE_CRYPTOAPI
+		if (CryptCreateHash(get_crypt_provider(), CALG_SHA_512, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+#elif defined TORRENT_USE_LIBCRYPTO
+		SHA512_Init(&m_context);
+#else
+		SHA512_init(&m_context);
+#endif
+	}
+
+	hasher512::hasher512(span<char const> data)
+		: hasher512()
+	{
+		update(data);
+	}
+
+	hasher512::hasher512(char const* data, int len)
+		: hasher512()
+	{
+		TORRENT_ASSERT(len > 0);
+		update({data, size_t(len)});
+	}
+
+#ifdef TORRENT_USE_LIBGCRYPT
+	hasher512::hasher512(hasher512 const& h)
+	{
+		gcry_md_copy(&m_context, h.m_context);
+	}
+
+	hasher512& hasher512::operator=(hasher512 const& h)
+	{
+		if (this == &h) return;
+		gcry_md_close(m_context);
+		gcry_md_copy(&m_context, h.m_context);
+		return *this;
+	}
+#elif TORRENT_USE_CRYPTOAPI
+	hasher512::hasher512(hasher512 const& h)
+	{
+		if (CryptDuplicateHash(h.m_context, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+	}
+
+	hasher512& hasher512::operator=(hasher512 const& h)
+	{
+		if (this == &h) return *this;
+		CryptDestroyHash(m_context);
+		if (CryptDuplicateHash(h.m_context, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+		return *this;
+	}
+#else
+	hasher512::hasher512(hasher512 const&) = default;
+	hasher512& hasher512::operator=(hasher512 const&) = default;
+#endif
+
+	hasher512& hasher512::update(char const* data, int len)
```

Comment:
this overload too. the reason I kept it around in `hasher` is because it's used in a bunch of places. I think